File: eng/ts/src/parser/declgrammar.ts
// The declarative grammar loader — the TS twin of eng/go/parser/
// declgrammar.go. Both engines load the SAME committed artifact
// (eng/go/parser/grammar.json; the Go side embeds it, this side reads
// it relatively, the SPEC_DIR precedent) so the grammar's structure —
// the token table and the rule-spec amendments — has one source of
// truth. Behavior that data cannot express binds by NAME from the
// host hook tables; an unknown name throws at parser construction.

import * as fs from 'node:fs'
import * as path from 'node:path'
import { fileURLToPath } from 'node:url'

const __dirname = path.dirname(fileURLToPath(import.meta.url))
const GRAMMAR_PATH = path.resolve(__dirname, '..', '..', '..', 'go', 'parser', 'grammar.json')

export interface DeclToken {
  name: string
  text?: string
}

export interface DeclAlt {
  s?: string[][]
  p?: string
  r?: string
  b?: number
  text?: string
  a?: string
  c?: string
}

export interface DeclEdit {
  rule: string
  op: string
  alts: DeclAlt[]
}

export interface DeclGrammar {
  schema: number
  tokens: DeclToken[]
  edits: DeclEdit[]
}

export interface DeclHooks {
  actions: Record<string, (r: any, ctx: any) => void>
  conds: Record<string, (r: any, ctx: any) => boolean>
}

let cached: DeclGrammar | undefined

export function loadDeclGrammar(): DeclGrammar {
  if (cached !== undefined) return cached
  const g = JSON.parse(fs.readFileSync(GRAMMAR_PATH, 'utf8')) as DeclGrammar
  if (g.schema !== 1) {
    throw new Error(`parser: grammar.json: unsupported schema ${g.schema}`)
  }
  cached = g
  return g
}

// fixedTokenOptions returns the fixed-token map for jsonic's options()
// (name '#OP' → source text), in artifact order.
export function fixedTokenOptions(g: DeclGrammar): Record<string, string> {
  const out: Record<string, string> = {}
  for (const tok of g.tokens) {
    if (tok.text !== undefined) out['#' + tok.name] = tok.text
  }
  return out
}

// resolveDeclTokens resolves every artifact token to its Tin — fixed
// tokens were pre-declared via options(), matcher-produced ones mint
// here, in artifact order.
export function resolveDeclTokens(j: any, g: DeclGrammar): Record<string, number> {
  const tins: Record<string, number> = {}
  for (const tok of g.tokens) {
    tins[tok.name] = j.token('#' + tok.name)
  }
  return tins
}

// applyDeclEdits applies every rule amendment in the artifact, binding
// named hooks from the tables — the exact twin of the Go loader. The
// helpers carry the host grammar's edit idioms plus its Text-marker
// constructor (the boxed-String newText).
export interface DeclHelpers {
  prependOpen: (rs: any, alts: any[]) => void
  prependClose: (rs: any, alts: any[]) => void
  setOpen: (rs: any, alts: any[]) => void
  setClose: (rs: any, alts: any[]) => void
  newText: (str: string) => unknown
}

export function applyDeclEdits(
  j: any,
  g: DeclGrammar,
  tins: Record<string, number>,
  hooks: DeclHooks,
  helpers: DeclHelpers,
): void {
  for (const edit of g.edits) {
    const alts = edit.alts.map((da) => buildDeclAlt(edit, da, tins, hooks, helpers))
    j.rule(edit.rule, (rs: any) => {
      switch (edit.op) {
        case 'prependOpen':
          helpers.prependOpen(rs, alts)
          break
        case 'appendOpen':
          rs.open(alts)
          break
        case 'setOpen':
          helpers.setOpen(rs, alts)
          break
        case 'prependClose':
          helpers.prependClose(rs, alts)
          break
        case 'appendClose':
          rs.close(alts)
          break
        case 'setClose':
          helpers.setClose(rs, alts)
          break
        default:
          throw new Error(`parser: grammar.json: unknown edit op ${edit.op} on rule ${edit.rule}`)
      }
    })
  }
}

function buildDeclAlt(
  edit: DeclEdit,
  da: DeclAlt,
  tins: Record<string, number>,
  hooks: DeclHooks,
  helpers: DeclHelpers,
): Record<string, unknown> {
  const alt: Record<string, unknown> = {}
  if (da.s !== undefined) {
    // The artifact's canonical shape is a sequence of token-alternative
    // groups; the TS jsonic flattens singleton groups.
    alt['s'] = da.s.map((group) => {
      const row = group.map((name) => {
        const tin = tins[name]
        if (tin === undefined) {
          throw new Error(`parser: grammar.json: unknown token ${name} in rule ${edit.rule}`)
        }
        return tin
      })
      return row.length === 1 ? row[0] : row
    })
  }
  if (da.p !== undefined) alt['p'] = da.p
  if (da.r !== undefined) alt['r'] = da.r
  if (da.b !== undefined) alt['b'] = da.b
  if (da.text !== undefined) {
    const marker = da.text
    alt['a'] = (r: any) => {
      r.node = helpers.newText(marker)
    }
  } else if (da.a !== undefined) {
    const fn = hooks.actions[da.a]
    if (fn === undefined) {
      throw new Error(`parser: grammar.json: unknown action hook ${da.a} in rule ${edit.rule}`)
    }
    alt['a'] = fn
  }
  if (da.c !== undefined) {
    const fn = hooks.conds[da.c]
    if (fn === undefined) {
      throw new Error(`parser: grammar.json: unknown condition hook ${da.c} in rule ${edit.rule}`)
    }
    alt['c'] = fn
  }
  return alt
}

File: eng/go/parser/declgrammar.go
package parser

// The declarative grammar loader. grammar.json is the single source of
// the boru grammar's STRUCTURE — the token table and the rule-spec
// amendments — shared verbatim with the TS twin (eng/ts reads the same
// file; this side embeds it). Behavior that data cannot express stays
// in the host language as NAMED hooks: `a` (action) and `c` (condition)
// names bind against the tables the caller passes to applyDeclEdits,
// and an unknown name is a panic at parser construction — the grammar
// file and the hook tables must agree exactly, on both engines.
//
// Migration is rule-by-rule: edits move from the setup* functions in
// grammar.go into grammar.json batch by batch, with the full parser
// suites (streamdump, corpora, the cross-engine differential) green
// after every batch. See design/ENG-COVERAGE-PARITY.0.md.

import (
	_ "embed"
	"encoding/json"
	"fmt"

	jsonic "github.com/tabnas/jsonic/go"
)

//go:embed grammar.json
var declGrammarJSON []byte

type declGrammar struct {
	Schema int         `json:"schema"`
	Tokens []declToken `json:"tokens"`
	Edits  []declEdit  `json:"edits"`
}

type declToken struct {
	Name string `json:"name"`
	Text string `json:"text"`
}

type declEdit struct {
	Rule string    `json:"rule"`
	Op   string    `json:"op"`
	Alts []declAlt `json:"alts"`
}

type declAlt struct {
	S    [][]string `json:"s"`
	P    string     `json:"p"`
	R    string     `json:"r"`
	B    int        `json:"b"`
	Text *string    `json:"text"`
	A    string     `json:"a"`
	C    string     `json:"c"`
}

// declHooks are the host-language bindings for the named behavior the
// grammar file references.
type declHooks struct {
	Actions map[string]func(*jsonic.Rule, *jsonic.Context)
	Conds   map[string]func(*jsonic.Rule, *jsonic.Context) bool
}

// loadDeclGrammar decodes the embedded artifact once per parse setup.
func loadDeclGrammar() declGrammar {
	var g declGrammar
	if err := json.Unmarshal(declGrammarJSON, &g); err != nil {
		panic(fmt.Sprintf("parser: grammar.json: %v", err))
	}
	if g.Schema != 1 {
		panic(fmt.Sprintf("parser: grammar.json: unsupported schema %d", g.Schema))
	}
	return g
}

// registerDeclTokens registers the artifact's token table in order —
// fixed tokens with their source text, matcher-produced tokens by name
// — and returns the name → Tin map the edits resolve against.
func registerDeclTokens(j *jsonic.Jsonic, g declGrammar) map[string]jsonic.Tin {
	tins := make(map[string]jsonic.Tin, len(g.Tokens))
	for _, tok := range g.Tokens {
		if tok.Text != "" {
			tins[tok.Name] = j.Token("#"+tok.Name, tok.Text)
		} else {
			tins[tok.Name] = j.Token("#" + tok.Name)
		}
	}
	return tins
}

// applyDeclEdits applies every rule amendment in the artifact, binding
// named hooks from the tables. Unknown rule ops, token names, or hook
// names panic: the artifact and the host tables are one contract.
func applyDeclEdits(j *jsonic.Jsonic, g declGrammar, tins map[string]jsonic.Tin, hooks declHooks) {
	for _, edit := range g.Edits {
		alts := make([]*jsonic.AltSpec, 0, len(edit.Alts))
		for _, da := range edit.Alts {
			alts = append(alts, buildDeclAlt(edit, da, tins, hooks))
		}
		op := edit.Op
		j.Rule(edit.Rule, func(rs *jsonic.RuleSpec, _ *jsonic.Parser) {
			switch op {
			case "prependOpen":
				prependOpen(rs, alts)
			case "appendOpen":
				rs.AddOpen(alts...)
			case "setOpen":
				setOpen(rs, alts)
			case "prependClose":
				prependClose(rs, alts)
			case "appendClose":
				rs.AddClose(alts...)
			case "setClose":
				setClose(rs, alts)
			default:
				panic(fmt.Sprintf("parser: grammar.json: unknown edit op %q on rule %q", op, edit.Rule))
			}
		})
	}
}

func buildDeclAlt(edit declEdit, da declAlt, tins map[string]jsonic.Tin, hooks declHooks) *jsonic.AltSpec {
	alt := &jsonic.AltSpec{P: da.P, R: da.R, B: da.B}
	if len(da.S) > 0 {
		s := make([][]jsonic.Tin, len(da.S))
		for i, group := range da.S {
			row := make([]jsonic.Tin, len(group))
			for k, name := range group {
				tin, ok := tins[name]
				if !ok {
					panic(fmt.Sprintf("parser: grammar.json: unknown token %q in rule %q", name, edit.Rule))
				}
				row[k] = tin
			}
			s[i] = row
		}
		alt.S = s
	}
	switch {
	case da.Text != nil:
		marker := *da.Text
		alt.A = func(r *jsonic.Rule, _ *jsonic.Context) {
			r.Node = jsonic.Text{Str: marker, Quote: ""}
		}
	case da.A != "":
		fn, ok := hooks.Actions[da.A]
		if !ok {
			panic(fmt.Sprintf("parser: grammar.json: unknown action hook %q in rule %q", da.A, edit.Rule))
		}
		alt.A = fn
	}
	if da.C != "" {
		fn, ok := hooks.Conds[da.C]
		if !ok {
			panic(fmt.Sprintf("parser: grammar.json: unknown condition hook %q in rule %q", da.C, edit.Rule))
		}
		alt.C = fn
	}
	return alt
}
